presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -392,6 +409,40 @@ public void enforceQueryMaxRunTimeLimits()
     }
 
     /**
+     * Enforce max driver number at the query level
+     */
+    public void enforceQueryMaxTasksLimits()
+    {
+        for (QueryExecution query : queries.values()) {
+            if (query.getQueryInfo().getState().isDone()) {
+                continue;
+            }
+            int queryMaxDrivers = SystemSessionProperties.getQueryMaxDrivers(query.getSession());
+            int totalDrivers = query.getQueryInfo().getQueryStats().getTotalDrivers();
+            if (queryMaxDrivers < totalDrivers) {
+                query.fail(new PrestoException(EXCEEDED_DRIVER_NUMBER_LIMIT, "Query exceeded maximum driver number limit of " + queryMaxDrivers));
```

Comment:
I really don't think limiting splits is a good idea. Many splits aren't related to data read (like splits for JOINs or final aggregations) and the number of those splits can vary based on local parallelism planning. It's entirely possible (and I see this in production queries) that your query has thousands of intermediate splits, and only hundreds of splits that actually read data. This could be changed to only count source splits, but even that seems unclear since splits can be of different sizes

I think having a configurable limit on data size read per query is totally reasonable though. We just need to figure out exactly what it would measure. I would suggest measuring the raw size of the data, so for Hive connector it would the bytes received from HDFS